macros/parsers/parserLinearRelation.pl

```diff
@@ -0,0 +1,379 @@
+	#
+	my $zero = 0 * $lN;
+	$ltype = $l->check_at($zero) ? 'eq' : 'ne';
+	$rtype = $r->check_at($zero) ? 'eq' : 'ne';
+
+	#
+	#  Reverse inequalities to favor lt, le over gt, ge
+	#
+	($lN, $ld, $ltype) = (-$lN, -$ld, $lrev) if $lrev;
+	($rN, $rd, $rtype) = (-$rN, -$rd, $rrev) if $rrev;
+
+	#
+	#  First, check if the type of inequality is the same.
+	#  Then check if the dividing (hyper)plane is the right one.
+	#
+	return 1 unless $ltype eq $rtype;
+
+	#
+	#  Are both 0?
+	#
+	if ($lN == $zero && $rN == $zero) {
+		my $ltruth = $l->check_at($zero);
+		my $rtruth = $r->check_at($zero);
+		return ($ltruth && $rtruth || !$ltruth && !$rtruth) ? 0 : 1;
+	}
+
+	# 'samedirection' is the second optional input to isParallel
+	return $lN <=> $rN
+		unless $lN->isParallel($rN, ($ltype ne 'eq' && $ltype ne 'ne'))
+		;    # inequalities require same direction which is the second optional input to isParallel
+	return $ld <=> $rd if $rd == 0 || $ld == 0;
+	return $rd * $lN <=> $ld * $rN;
+}
+
+sub cmp_class {'a Linear Relation'}
+sub showClass { shift->cmp_class }
+
+sub cmp_defaults { (
+	Value::Real::cmp_defaults(shift),
+	ignoreInfinity => 0,    # report infinity as an error
+) }
+
+#
+#  Only compare two relations
+#
+sub typeMatch {
+	my ($self, $other, $ans) = @_;
+	return ref($other) && $other->type eq 'Relation' unless ref($self);
+	return ref($other) && $self->type eq $other->type;
+}
+
+sub string {
+	my $self = shift;
+	if (!$self->getFlag(standardForm) && defined $self->{original_formula}) {
+		return $self->{original_formula};
+	} else {
+		return $self->SUPER::string;
+	}
+}
+
+sub TeX {
+	my $self = shift;
+	if (!$self->getFlag(standardForm) && defined $self->{original_formula_latex}) {
+		return $self->{original_formula_latex};
+	} else {
+		return $self->SUPER::TeX;
+	}
+}
+
+sub check_at {
+	my ($self, $point) = @_;
+	if (ref($point) ne 'ARRAY') {
+		$self->Error("check_at argument must be an array reference or MathObject Point or Vector.")
+			unless ($point->type eq 'Point' || $point->type eq 'Vector');
+	}
+	my $context = $self->context;
+	$point = $context->Package("Vector")->make($context, @{$point}) if (ref($point) eq 'ARRAY');
+	my @variables = main::lex_sort(keys(%{ $context->{variables} }));
+	my $n         = @variables;
+	$self->Error("The context for this linear relation has $n variables: "
+			. join(', ', @variables)
+			. ", so a point to check at must also have $n entries")
+		unless ($n == $point->value);
+	return $self->eval(map { $variables[$_] => ($point->value)[$_] } (0 .. $#variables));
+}
+
+#
+#  We subclass BOP::equality so that we can assign a type using _check and
+#  override the _eval method for relation operators
+#
+
+package LinearRelation::inequality;
+our @ISA = qw(Parser::BOP::equality);
+
+sub _check {
+	my $self = shift;
+	$self->SUPER::_check;
+	$self->{type} = Value::Type('Relation', 1);
+}
+
+sub _eval {
+	my $self = shift;
+	return &{ $self->{def}{eval} }(@_);
```

Comment:
After writing this, I remembered that `_eval` is supposed to return a MathObject, though it probably doesn't matter in this case.  If you want to, you could use

``` perl
	my $context = $self->context;
	return $context->Package("Real")->new($context, &{ $self->{def}{eval} }(@_) ? 1 : 0);
```

or something like that.  But again, it probably doesn't really matter here, so might not be worth it.

Reply:
Thanks for all your feedback. I have applied these changes. I was under the false impression that `$context` was a copy of the current context instead of the actual current context.